pkg/multiclustermanager/routes.go

```diff
@@ -93,11 +93,20 @@ func router(ctx context.Context, localClusterEnabled bool, tunnelAuthorizer *mcm
 	unauthed.PathPrefix("/v3-public").Handler(publicAPI)
 
 	// Authenticated routes
-	authed := mux.NewRouter()
-	authed.UseEncodedPath()
 	impersonatingAuth := auth.ToMiddleware(requests.NewImpersonatingAuth(sar.NewSubjectAccessReview(clusterManager)))
+	saAuth := auth.ToMiddleware(requests.NewServiceAccountAuth(scaledContext, clustermanager.ToRESTConfig))
 	accessControlHandler := rbac.NewAccessControlHandler()
 
+	saauthed := mux.NewRouter()
+	saauthed.UseEncodedPath()
+	saauthed.PathPrefix("/k8s/clusters/{clusterID}").Handler(k8sProxy)
+	saauthed.Use(mux.MiddlewareFunc(saAuth.Chain(impersonatingAuth)))
```

Comment:
I know this is a very delicate part, but I guess this can be dropped and chained as the other middlewares:

```suggestion
	saauthed.Use(mux.MiddlewareFunc(saAuth))
	saauthed.Use(mux.MiddlewareFunc(impersonatingAuth))
```

We can also reuse the same "authentication chain" in the following routers, maybe it's cleaner.
Something like this should work but I haven't tested it:

```go

authenticationMiddlewares := []mux.MiddlewareFunc{
	mux.MiddlewareFunc(impersonatingAuth),
	mux.MiddlewareFunc(accessControlHandler),
	requests.NewAuthenticatedFilter,
}

// [...]
// sauthed router
saauthed.Use(mux.MiddlewareFunc(saAuth))
saauthed.Use(authenticationMiddlewares...)

// [...]
// authed router
authed := mux.NewRouter()
authed.UseEncodedPath()
authed.Use(authenticationMiddlewares...)

// [...]
// metricsAuthed router
metricsAuthed.Use(mux.MiddlewareFunc(tokenReviewAuth))
metricsAuthed.Use(authenticationMiddlewares...)
```

Reply:
I know that @pmatseykanets did some experimentation around this.  Any thoughts, Peter?